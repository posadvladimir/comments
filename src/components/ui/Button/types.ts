
export type TButtonProps = {
    onClick: React.MouseEventHandler<HTMLButtonElement>;
    disabled: boolean;
    text: string | number;
    isLoading?: boolean;
}