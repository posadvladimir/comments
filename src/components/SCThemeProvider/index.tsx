
import { ThemeProvider } from 'styled-components';

import type { TChildrenProps } from 'src/ts/types/children';

import theme from 'src/styles/theme';

const SCThemeProvider: React.FC<TChildrenProps> = ({ children }) => (
    <ThemeProvider theme={theme}>
        {children}
    </ThemeProvider>
);

export default SCThemeProvider;