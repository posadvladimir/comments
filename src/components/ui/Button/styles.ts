
import { styled } from 'styled-components';

export const Button = styled.button`
    width: 234px;
    background: #313439;
    border-radius: 4px;
    padding: 8px 0;
    transition: background .1s ease;
    position: relative;

    &:hover {
        background: #1f2023;
    }

    &:active {
        background: #17171a;
    }

    &[disabled] {
        cursor: auto;
        background: #494949;
        color: #8a8a8a;
    }
`;

export const ButtonText = styled.span`
    color: inherit;
    line-height:  ${props => props.theme.default.lineHeight}px;
`;