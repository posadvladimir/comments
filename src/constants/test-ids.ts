
const LOADER_TEST_ID = 'loader-test-id';
const BUTTON_TEST_ID = 'button-test-id';
const LIKE_COUNTER_TEST_ID = 'like-counter-test-id';
const LIKE_COUNTER_BUTTON_TEST_ID = 'like-counter-button-test-id';
const LIKE_COUNTER_TOTAL_TEST_ID = 'like-counter-total-test-id';
const COMMENT_TEST_ID = 'comment-test-id';
const ERROR_TEST_ID = 'error-test-id';

export {
    LOADER_TEST_ID,
    BUTTON_TEST_ID,
    LIKE_COUNTER_TEST_ID,
    LIKE_COUNTER_BUTTON_TEST_ID,
    COMMENT_TEST_ID,
    LIKE_COUNTER_TOTAL_TEST_ID,
    ERROR_TEST_ID,
};