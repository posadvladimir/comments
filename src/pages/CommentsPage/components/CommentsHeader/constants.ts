
import type { TDeclensionArray } from '../../common/ts/types/declensions';

export const COMMENT_DECLENSIONS: TDeclensionArray = [
    'комментарий',
    'комментария',
    'комментариев'
];