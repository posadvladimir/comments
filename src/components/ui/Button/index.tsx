
import type { TButtonProps } from './types';

import { BUTTON_TEST_ID } from 'src/constants/test-ids';

import * as S from './styles';

import Loader from 'src/components/ui/Loader';

const Button: React.FC<TButtonProps> = ({
    onClick,
    disabled,
    text,
    isLoading = false,
}) => (
    <S.Button
        onClick={onClick}
        disabled={disabled}
        data-testid={BUTTON_TEST_ID}
    >
        <S.ButtonText>
            {text}
        </S.ButtonText>

        {isLoading && <Loader />}
    </S.Button>
);

export default Button;