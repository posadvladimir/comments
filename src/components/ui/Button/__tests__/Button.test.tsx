
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { LOADER_TEST_ID, BUTTON_TEST_ID } from 'src/constants/test-ids';

import { runSnapshotTest } from 'src/common/tests/snapshot';

import SCThemeProvider from 'src/components/SCThemeProvider';
import Button from '..';

const onClickMock = jest.fn();

const getButtonComponent = ({
    disabled = false,
    isLoading = false,
}) => (
    <SCThemeProvider>
        <Button
            onClick={onClickMock}
            disabled={disabled}
            text="Click me"
            isLoading={isLoading}
        />
    </SCThemeProvider>
);

const renderButton = ({
    disabled = false,
    isLoading = false,
}): void => {
    render(getButtonComponent({ disabled, isLoading }));
};

describe('Button', () => {
    test('вызывается функция onClick при клике на кнопку', () => {
        renderButton({});
    
        const button = screen.getByTestId(BUTTON_TEST_ID);
        userEvent.click(button);
    
        expect(onClickMock).toHaveBeenCalledTimes(1);
    });

    test('отображается лоадер при isLoading равном true', () => {
        renderButton({ isLoading: true });

        expect(screen.getByTestId(LOADER_TEST_ID)).toBeInTheDocument();
    });

    test('блокируется кнопка при disabled равном true', () => {
        renderButton({ disabled: true });

        expect(screen.getByTestId(BUTTON_TEST_ID)).toBeDisabled();
    });

    runSnapshotTest(getButtonComponent({}));
});