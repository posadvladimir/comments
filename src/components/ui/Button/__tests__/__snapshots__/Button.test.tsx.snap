// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button соотвествует снепшоту 1`] = `
<button
  className="sc-bcPKhP jYRdSL"
  data-testid="button-test-id"
  disabled={false}
  onClick={[MockFunction]}
>
  <span
    className="sc-grXZZQ huinXV"
  >
    Click me
  </span>
</button>
`;
